### DISCUSSION TITLE
How to visualize an "ab" slice of CIE Lab with corresponding "L"?

### DISCUSSION THREAD

**Discussion by hminle** [colour-science/colour#858] [Q&A] [answered]
Hi,

Do you guys have any idea on how to visualize an "ab" plane of CIE Lab with corresponding "L"?
Like what've been shown in this webpage: http://web.mit.edu/kenta/www/one/color-lab/dwvcvceg/color.html

Best,

**Comment by thomasmansencal** [comment, marked as answer]
Hi @hminle,

We do not have anything readily available for that and while it is not easy to do that for arbitrary gamuts as it requires mesh geometry intersection, it is trivial to do for RGB colourspaces:

```python
from colour.colorimetry import CCS_ILLUMINANTS, SDS_ILLUMINANTS
from colour.geometry import primitive_cube
from colour.models import Lab_to_XYZ, XYZ_to_Lab, XYZ_to_sRGB, sRGB_to_XYZ
from colour.plotting import XYZ_to_plotting_colourspace, artist
from colour.utilities import as_int, full, orient, tstack
from colour.volume import is_within_visible_spectrum, solid_RoschMacAdam

illuminant = 'D65'
sd_n = SDS_ILLUMINANTS[illuminant]
xy_n = CCS_ILLUMINANTS['CIE 1931 2 Degree Standard Observer'][
    illuminant]

vertices, faces, outline = primitive_cube(1, 1, 1, 32, 32, 32)
XYZ_vertices = sRGB_to_XYZ(
    vertices['position'] + 0.5,
    apply_cctf_decoding=False)
Lab_vertices = XYZ_to_Lab(XYZ_vertices, xy_n)

extent = np.max(
    np.abs([np.min(Lab_vertices[..., 1:]),
            np.max(Lab_vertices[..., 1:])]))
extent = np.around(extent * 1.1 / 100, 1) * 100

samples = 2048
L = 50

ii, jj = np.meshgrid(
    np.linspace(-extent, extent, samples), np.linspace(extent, -extent,
                                                       samples))
ij = tstack([ii, jj])

Lab_slice = full([samples, samples, 3], L)
Lab_slice[..., 1:] = ij

normal = [1, 0, 0]

XYZ_slice = Lab_to_XYZ(Lab_slice, xy_n)
RGB_slice = XYZ_to_plotting_colourspace(XYZ_slice, xy_n)
XYZ_mask = XYZ_to_sRGB(XYZ_slice, apply_cctf_encoding=False)
RGB_slice[np.any(XYZ_mask < 0, axis=-1)] = 1
RGB_slice[np.any(XYZ_mask > 1, axis=-1)] = 1

figure, axes = artist()
image = axes.imshow(
    RGB_slice,
    interpolation='bilinear',
    extent=(-extent, extent, -extent, extent),
    clip_path=None);
```

![image](https://user-images.githubusercontent.com/99779/131245281-64a15e04-8260-4850-82c7-9c4fb6a46ebb.png)

**Reply by hminle** [reply]
Great, thank you so much @KelSolaar 
I will try it out

**Reply by hminle** [reply]
Hi @KelSolaar , what is your colour version?
Since I cannot import some function.
![image](https://user-images.githubusercontent.com/16201681/131259516-8f1dd770-286d-482c-b0b6-5f9cd34ed6e5.png)

**Reply by hminle** [reply]
Hi @KelSolaar ,

I have some questions:
1. What is `primitive_cube`? I see the documentation, but it's too short. I cannot get it.
As I guess, premitive cube returns sRGB gamut color space. But then you take the vertices position and plus 0.5 `vertices['position'] + 0.5` ? Don't know why `0.5` here?

2. After you get the `Lab_vertices` of sRGB vertices, you create a meshgrid of `ab` channel with L=50. Then you create a `Lab_slice` according to L=50. Here why you don't visualize `Lab_slice` but you need to convert it back to `RGB_slice` using `XYZ_to_plotting_colourspace` ? (I cannot find the documentation of `XYZ_to_plotting_colourspace`, but as I read the source code, it will transform XYZ to RGB)

3. Why we need to have `XYZ_mask` and set `RGB_slice` to 1 for every XYZ_mask > 1 and < 0? Is it for out of gamut?

4. I see you set `normal` variable, but don't use it?

5. This is apply for sRGB, what if I want to visualize AdobeRGB, what should I change? Should I need to change this line of code?
```python
XYZ_vertices = sRGB_to_XYZ(
    vertices['position'] + 0.5,
    apply_cctf_decoding=False)
```
to
```python
adobeRGB = RGB_COLOURSPACES['Adobe RGB']
XYZ_vertices = RGB_to_XYZ(
        vertices['position'] + 0.5,,
        adobeRGB.whitepoint,
        illuminant=adobeRGB.whitepoint,
        matrix_RGB_to_XYZ=adobeRGB.matrix_RGB_to_XYZ,
        chromatic_adaptation_transform='CAT02,
        cctf_decoding=None,
    )
```

Thank you in advance

**Reply by thomasmansencal** [reply]
Hi,

Sorry, the code is not super clean because I was experimenting with geometry an `trimesh` (that we use in a feature branch) on the side. I'm pretty much always using the latest of `Colour`, at anytime:

```
===============================================================================
*                                                                             *
*   Interpreter :                                                             *
*       python : 3.7.11 (default, Jul  3 2021, 18:01:19)                      *
*                [GCC 7.5.0]                                                  *
*                                                                             *
*   colour-science.org :                                                      *
*       colour : v0.3.16-366-gf1f00c14                                        *
*                                                                             *
*   Runtime :                                                                 *
*       imageio : 2.4.1                                                       *
*       matplotlib : 3.2.2                                                    *
*       networkx : 2.6.2                                                      *
*       numpy : 1.19.5                                                        *
*       pandas : 1.1.5                                                        *
*       scipy : 1.4.1                                                         *
*       tqdm : 4.62.0                                                         *
*                                                                             *
===============================================================================
```

1. `primitive_cube` is a definition that returns the geometrical components for a cube, i.e., the vertices positions, uv, normal and colour, the faces indices and a nice outline.
2. Here, I could have used `XYZ_to_sRGB` but ultimately, I will roll those functions cleanly in Colour, there is related issues here: https://github.com/colour-science/colour/issues/647
3. Yes, you want to capture the OOG colours to slice the solid, try to comment the `RGB_slice[np.any(XYZ_mask < 0, axis=-1)] = 1` and `RGB_slice[np.any(XYZ_mask > 1, axis=-1)] = 1` statements one after each other and you will see what is happening.
4. It is because I have the code with the `trimesh` variant here and it is very much WIP: https://colab.research.google.com/drive/1SGk3DWzWanLAbdbl_DBzgeikgM3lxddH?usp=sharing
5. That would be the idea, but there are two things at play, there is the plotting colourspace and there is the gamut colourspace which is different, my current code does not separate them super cleanly and I think you would have issues with it right now without some tweaks. I will clean that up a little bit.

**Reply by hminle** [reply]
Thank you @KelSolaar , I got your points.

> there is the plotting colourspace and there is the gamut colourspace which is different

This is a clever trick. I cannot think of it at the moment.

**Reply by thomasmansencal** [reply]
I spent a bit of time refining the code for that, I might have something by the end of week that could be included in Colour. Stay tuned!

**Reply by hminle** [reply]
Thank you @KelSolaar :)

**Reply by hminle** [reply]
Hi @KelSolaar ,

In this case, how you can visualize the slice of another gamut, like "ProPhoto RGB"?
I tried to visualize both Lab slice of "sRGB" and "ProPhoto RGB" on the same figure.
But got the weird result.
![image](https://user-images.githubusercontent.com/16201681/134062375-8882d42d-648b-4129-b8e0-f12312847163.png)

The ProPhoto region is blue
The sRGB region is yellow.

But I think I made a mistake somewhere, since the ProPhoto region is smaller than sRGB.

**Reply by Myndex** [reply]
Hi @hminle 

> But I think I made a mistake somewhere, since the ProPhoto region is smaller than sRGB.

So I am going to guess that the vertical axis is ±b*, and the excursion you see in blue is common when you are mixing illuminants D50 and D65 without the appropriate transform.

sRGB is D65, and ProPhoto is D50. They don't mix (not directly) you need to adapt one or the other, and your LAB space needs to be adapted to that same whitepoint as well.

This "looks like" (taking a guess) that your LAB space is at D50, as well as ProPhoto, and that you brought in the sRGB without first doing a Bradford (or other) chromatic adaptation transform. If you want your LAB space to be in ProPhoto's D50, then you need to transform sRGB to D50. D50 is the illuminant used for the PCS in v2 and v4 ICC profiles.

Alternately, you could adapt your ProPhoto to D65, and use LAB with D65, and sRGB then would remain as D65.

**Reply by hminle** [reply]
Hi @Myndex ,

I convert everything to D50 illuminant. 
Belows are my converting function.
The vertical axis is "a" channel, the horizontal is "b"

```python
ILLUMINANT_D50 = colour.CCS_ILLUMINANTS['CIE 1931 2 Degree Standard Observer']['D50']

def srgb_to_Lab_D50(srgb):
    srgb_colourspace = colour.models.RGB_COLOURSPACE_sRGB.chromatically_adapt(ILLUMINANT_D50)

    XYZ = colour.models.RGB_to_XYZ(img_srgb.reshape(-1,3), 
                                         srgb_colourspace.whitepoint, 
                                         srgb_colourspace.whitepoint,
                                         srgb_colourspace.matrix_RGB_to_XYZ)
    Lab_srgb = colour.XYZ_to_Lab(XYZ, 
                                 illuminant=ILLUMINANT_D50)

    return Lab_srgb

def prop_to_Lab_D50(prop):
    prophotorgb_colourspace = colour.models.RGB_COLOURSPACE_PROPHOTO_RGB

    XYZ = colour.models.RGB_to_XYZ(img_prop.reshape(-1,3), 
                                 prophotorgb_colourspace.whitepoint, 
                                 prophotorgb_colourspace.whitepoint,
                                 prophotorgb_colourspace.matrix_RGB_to_XYZ)
    Lab_prop = colour.XYZ_to_Lab(XYZ, 
                                 illuminant=ILLUMINANT_D50)

    return Lab_prop

def lab_to_prop_D50(Lab):
    prophotorgb_colourspace = colour.models.RGB_COLOURSPACE_PROPHOTO_RGB
    XYZ = colour.Lab_to_XYZ(Lab, 
                            illuminant=ILLUMINANT_D50)
    
    img = colour.models.XYZ_to_RGB(XYZ, 
                 prophotorgb_colourspace.whitepoint, 
                 prophotorgb_colourspace.whitepoint,
                 prophotorgb_colourspace.matrix_XYZ_to_RGB)
    return img

def lab_to_srgb_D50(Lab):
    srgb_colourspace = colour.models.RGB_COLOURSPACE_sRGB.chromatically_adapt(ILLUMINANT_D50)
    XYZ = colour.Lab_to_XYZ(Lab, 
                            illuminant=ILLUMINANT_D50)
    
    img = colour.models.XYZ_to_RGB(XYZ, 
                 srgb_colourspace.whitepoint, 
                 srgb_colourspace.whitepoint,
                 srgb_colourspace.matrix_XYZ_to_RGB)

    return img
```

**Comment by Myndex** [comment]
Hi @hminle 

Well if the vertical is ±a* then sRGB is exceeding ProPhoto in green, that's crazy. 

Unfortunately I don't know Ken's code well enough to suggest a fix. But I'd suggest looking someplace where ProPhoto is being clipped.

**Comment by hminle** [comment]
Hi @KelSolaar ,

Is there anyway I can visualize 2 colour spaces (sRGB and ProPhoto RGB) in one Lab slice?
Maybe don't need `RGB_slice`, but only scatter points in `ab space`

**Reply by hminle** [reply]
Oh I got it.

Here is my code:

```python

def get_Lab_slice(colourspace, value_L=50):
    samples = 256
    colourspace_Lab_limits = colour.volume.RGB_colourspace_limits(colourspace)
    ii, jj = np.meshgrid(
        np.linspace(colourspace_Lab_limits[1,0], colourspace_Lab_limits[1,1], samples), 
        np.linspace(colourspace_Lab_limits[2,0], colourspace_Lab_limits[2,1], samples))
    ij = tstack([ii, jj])

    Lab_slice = full([samples, samples, 3], value_L)
    Lab_slice[..., 1:] = ij
    XYZ_slice = Lab_to_XYZ(Lab_slice, colourspace.whitepoint)

    RGB_mask = XYZ_to_RGB(XYZ_slice,
                          colourspace.whitepoint,
                          colourspace.whitepoint,
                          colourspace.matrix_XYZ_to_RGB,)
    oog_mask = ((RGB_mask < 0) | (RGB_mask > 1)).any(axis=2)
    return Lab_slice[~oog_mask]

ILLUMINANT_D65 = colour.CCS_ILLUMINANTS['CIE 1931 2 Degree Standard Observer']['D65']
prophotorgb = colour.models.PROPHOTO_RGB_COLOURSPACE.chromatically_adapt(ILLUMINANT_D65)
srgb = colour.models.sRGB_COLOURSPACE

prop_Lab_slice = get_Lab_slice(prophotorgb)
srgb_Lab_slice = get_Lab_slice(srgb)

plt.scatter(prop_Lab_slice[:, 1], prop_Lab_slice[:, 2], c='b')
plt.scatter(srgb_Lab_slice[:, 1], srgb_Lab_slice[:, 2], c='r')
plt.xlim([-200, 200])
plt.ylim([-200, 200])
plt.show()
```

![image](https://user-images.githubusercontent.com/16201681/135730885-52eb2ab0-8976-4060-88de-38f38d4a1e94.png)

**Reply by thomasmansencal** [reply]
Hi @hminle,

If you use the [feature/gamut_section](https://github.com/colour-science/colour/tree/feature/gamut_section) branch, the following code produces the attached image:

```python
figure, axes = plot_visible_spectrum_section(
    model='CIE Lab',
    origin=0.5,
    section_colours='RGB',
    contour_colours='RGB',
    section_opacity=0.15,
    standalone=False)

figure, axes = plot_RGB_colourspace_section(
    colourspace=colour.models.RGB_COLOURSPACE_PROPHOTO_RGB.chromatically_adapt(
        colour.models.RGB_COLOURSPACE_sRGB.whitepoint),
    model='CIE Lab',
    origin=0.5,
    section_colours=[1, 0.25, 0],
    section_opacity=0.15,
    axes=axes,
    standalone=False)

bounding_box = [
    axes.get_xlim()[0],
    axes.get_xlim()[1],
    axes.get_ylim()[0],
    axes.get_ylim()[1]
]
plot_RGB_colourspace_section(
    colourspace='sRGB',
    model='CIE Lab',
    origin=0.5,
    section_colours=[0, 0.25, 1],
    section_opacity=0.15,
    axes=axes,
    bounding_box=bounding_box,
    title=(
        'sRGB, ProPhoto RGB Sections - '
        'L = 50% - '
        'CIE Lab (Domain-Range Scale: 1)'))

```

![image](https://user-images.githubusercontent.com/99779/135736605-0d51f57a-4199-4c47-b519-bd6d8b30d3b2.png)

Note that almost all the colour models are normalised to be comparable [as per the documentation](https://colour.readthedocs.io/en/develop/basics.html#scale-1).

Cheers,

Thomas

**Reply by hminle** [reply]
Hi @KelSolaar ,

Thank for the info.
By the way, I don't understand "the colour models are normalised"

> Note that almost all the colour models are normalised to be comparable as per the documentation.

Now what if I want to visualize ProPhoto or sRGB image pixels on this image.
What should I do? How can I normalize the image pixels?
What if I just use ‘Reference’ default domain-range scale to visualize both image pixels and colour space boundary in the same CIE Lab figure?

**Reply by thomasmansencal** [reply]
The Reference domain would not work here because the conversion uses `colour.convert`, I can probably add a post-scale though, I will think about it.

**Reply by thomasmansencal** [reply]
Same code than above in the latest from the [feature/gamut_section](https://github.com/colour-science/colour/tree/feature/gamut_section) branch: 
![image](https://user-images.githubusercontent.com/99779/135741445-46081a7e-301d-4bc0-ad6c-f4043f1c10eb.png)

**Reply by hminle** [reply]
Wow, thank @KelSolaar
It seems that you add the code for normalizing axis or not? right?
https://github.com/colour-science/colour/commit/5341ae0478ec8859afd6de09cb6f209b769d1683#diff-08b38bdbe17c18a6b9091c4d4daaf173d0e8639f54a83c12ab28c031c7bcb145R209

so if I want to visualize image pixels, I need to use `convert` to convert image colourspace to reference colourspace, right?

Or another way is to multiply 100 for both image pixels and colour space?

**Reply by thomasmansencal** [reply]
So, the normalisation code you linked serves a slightly different purpose to that of the domain-range scale, the idea is that some of the colour models, e.g. ICtCp, have absolute scales so it becomes really hard to set 50% lightness with them, it might actually be something like 0.027 or alike. If `normalise` is on, and you use 0.5 for the origin, the section will be in the middle of the data along the chosen axis.

When you say image pixels, would you need to add a scatter point of image data on-top of the section(s)?

**Reply by thomasmansencal** [reply]
@hminle : The code has been merged in the develop branch!

**Reply by hminle** [reply]
Greatt, thank @KelSolaar . I will try it out.

**Reply by hminle** [reply]
> So, the normalisation code you linked serves a slightly different purpose to that of the domain-range scale, the idea is that some of the colour models, e.g. ICtCp, have absolute scales so it becomes really hard to set 50% lightness with them, it might actually be something like 0.027 or alike. If `normalise` is on, and you use 0.5 for the origin, the section will be in the middle of the data along the chosen axis.
> 
> When you say image pixels, would you need to add a scatter point of image data on-top of the section(s)?

Hi @KelSolaar , yes. I want to add scatter point of image data on--top of the sections. Just want to see out-of-gamut pixels

**Reply by thomasmansencal** [reply]
Try something like that maybe: https://colab.research.google.com/drive/1bljhLAfzm-36GQfjALYrQdHuf7fNYTaw#scrollTo=oAwadLyX1GQB

**Reply by hminle** [reply]
Thank @KelSolaar , I've just requested an access to your link.
